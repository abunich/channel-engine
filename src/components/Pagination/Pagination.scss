@import "src/utils/variables";

.pagination {
    display: flex;
    list-style-type: none;
}