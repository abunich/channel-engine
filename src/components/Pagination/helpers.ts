export const PREV_VALUE = "<";
export const NEXT_VALUE = ">";
export const MAX_PAGINATION_ELEMENTS = 3;

export const getPaginationItemValue = (num: number, nextArrowValue: number) =>
  num === 0 ? PREV_VALUE : num === nextArrowValue ? NEXT_VALUE : num;

const getFirstElement = (page: number, pages: number, numberElements: number) => {
  const gap = Math.ceil(numberElements / 2 - 1);

  const isRightGapMoreThanPages = page + gap > pages;
  if (isRightGapMoreThanPages) {
    const result = pages - numberElements;
    return result > 0 ? result + 1 : 0;
  }

  const isLeftGapLessThanFirstPage = page - gap < 1;
  if (isLeftGapLessThanFirstPage) {
    return 1;
  }

  return page - gap;
};

export const getArrayOfPagination = (page: number, pages: number, limit: number): number[] => {
  const [numberElements, firstElement] = pages > limit ? [limit, getFirstElement(page, pages, limit)] : [pages, 1];

  const array = new Array(numberElements).fill(0).map((_, index) => {
    return firstElement + index;
  });

  return [0, ...array, pages + 1];
};
