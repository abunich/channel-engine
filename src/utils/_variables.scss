$primary-color: #0c317a;
$shadow: rgba(0, 0, 0, 0.1) 0px 4px 12px;
$border-color: #E8E8E8;
$white-color: #FFFFFF;