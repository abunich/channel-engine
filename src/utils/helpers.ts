import { DateTime } from "luxon";

export const getRequiredTimeFormat = (date: string) => {
  const newDate = DateTime.fromISO(date).toFormat("MM-dd-yyyy hh:mm");

  return newDate;
};
