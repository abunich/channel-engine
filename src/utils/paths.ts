export const OVERVIEW_URL = "/";

export const DETAILS_URL = "/details";