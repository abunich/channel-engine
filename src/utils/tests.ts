import renderer from "react-test-renderer";

export const getTestingComponent = (component: JSX.Element) => {
    return renderer.create(component).toJSON();
}

export const toMatchSnapshot = (component: JSX.Element): void  => {
  const tree = getTestingComponent(component);

  expect(tree).toMatchSnapshot();
}
