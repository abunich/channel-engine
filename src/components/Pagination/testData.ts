import { Pagination } from "./Pagination";

export const withoutData: Pagination.Props = {
    page: 1,
    totalCount: 0,
    limit: 3,
    changePage: () => {}
}

export const firstPageData: Pagination.Props = {
    page: 1,
    totalCount: 25,
    limit: 3,
    changePage: () => {}
}

export const lastPageData: Pagination.Props = {
    page: 5,
    totalCount: 25,
    limit: 5,
    maxDisplayedPages: 5,
    changePage: () => {}
}