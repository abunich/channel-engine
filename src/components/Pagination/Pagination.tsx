import React from "react";
import { PaginationItem } from "./components/PaginationItem/PaginationItem";
import { getArrayOfPagination, MAX_PAGINATION_ELEMENTS } from "./helpers";
import "./Pagination.scss";

export declare namespace Pagination {
  interface Props {
    page: number;
    totalCount: number;
    limit: number;
    maxDisplayedPages?: number;
    changePage: (page: number) => void;
  }
}

export const Pagination: React.FC<Pagination.Props> = ({
  page,
  totalCount,
  limit,
  maxDisplayedPages = MAX_PAGINATION_ELEMENTS,
  changePage,
}) => {
  const pages = Math.ceil(totalCount / limit);

  if (!pages) {
    return null;
  }

  const pagesButtonNumber =
    pages > 1 ? getArrayOfPagination(page, pages, maxDisplayedPages) : [pages];
  const nextArrowValue = pages + 1;

  const selectPage = (value: number) => {
    const newPage = value === 0 ? page - 1 : value === nextArrowValue ? page + 1 : value;

    changePage(newPage);
  };

  const onClick = (e: React.MouseEvent<HTMLLIElement>) => {
    selectPage(e.currentTarget.value);
  };

  const paginationItems = pagesButtonNumber.map((paginationNumber) => (
    <PaginationItem
      key={paginationNumber}
      page={page}
      pages={pages}
      value={paginationNumber}
      onClick={onClick}
    />
  ));

  return <ul className="pagination">{paginationItems}</ul>;
};
