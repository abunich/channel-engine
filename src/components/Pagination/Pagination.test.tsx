/**
 *  @jest-environment jsdom
 */
import React from "react";
import { render, screen } from "@testing-library/react";
import { toMatchSnapshot } from "src/utils/tests";
import { Pagination } from "./Pagination";
import { withoutData, firstPageData, lastPageData } from "./testData";

const activeClass = "pagination__item pagination__item_active";
const disabledClass = "pagination__item pagination__item_disabled";

describe("<Pagination />", () => {
  it("Render Pagination with no data", () => {
    toMatchSnapshot(<Pagination {...withoutData} />);
  });

  it("Render Pagination with active first page", () => {
    const changePage = jest.fn();

    render(<Pagination {...firstPageData} changePage={changePage} />);

    const firstButton = screen.queryByTestId("pagination-item-1");

    expect(firstButton?.className).toBe(activeClass);

    firstButton?.click();

    expect(changePage).toBeCalledTimes(0);

    const prevButton = screen.queryByTestId("pagination-item-0");

    expect(prevButton?.className).toBe(disabledClass);

    prevButton?.click();

    expect(changePage).toBeCalledTimes(0);

    const secondButton = screen.queryByTestId("pagination-item-2");

    secondButton?.click();

    expect(changePage).toBeCalledTimes(1);

    toMatchSnapshot(<Pagination {...firstPageData} />);
  });

  it("Render Pagination with active last page", () => {
    const changePage = jest.fn();

    render(<Pagination {...lastPageData} changePage={changePage} />);

    const lastButton = screen.queryByTestId("pagination-item-5");

    expect(lastButton?.className).toBe(activeClass);

    lastButton?.click();

    expect(changePage).toBeCalledTimes(0);

    const nextButton = screen.queryByTestId("pagination-item-6");

    expect(nextButton?.className).toBe(disabledClass);

    nextButton?.click();

    expect(changePage).toBeCalledTimes(0);

    const preLastButton = screen.queryByTestId("pagination-item-4");

    preLastButton?.click();

    expect(changePage).toBeCalledTimes(1);

    toMatchSnapshot(<Pagination {...lastPageData} />);
  });
});
