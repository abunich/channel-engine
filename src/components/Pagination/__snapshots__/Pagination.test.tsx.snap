// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> Render Pagination with active first page 1`] = `
<ul
  className="pagination"
>
  <li
    className="pagination__item pagination__item_disabled"
    data-testid="pagination-item-0"
    onClick={[Function]}
    value={0}
  >
    &lt;
  </li>
  <li
    className="pagination__item pagination__item_active"
    data-testid="pagination-item-1"
    onClick={[Function]}
    value={1}
  >
    1
  </li>
  <li
    className="pagination__item"
    data-testid="pagination-item-2"
    onClick={[Function]}
    value={2}
  >
    2
  </li>
  <li
    className="pagination__item"
    data-testid="pagination-item-3"
    onClick={[Function]}
    value={3}
  >
    3
  </li>
  <li
    className="pagination__item"
    data-testid="pagination-item-10"
    onClick={[Function]}
    value={10}
  >
    &gt;
  </li>
</ul>
`;

exports[`<Pagination /> Render Pagination with active last page 1`] = `
<ul
  className="pagination"
>
  <li
    className="pagination__item"
    data-testid="pagination-item-0"
    onClick={[Function]}
    value={0}
  >
    &lt;
  </li>
  <li
    className="pagination__item"
    data-testid="pagination-item-1"
    onClick={[Function]}
    value={1}
  >
    1
  </li>
  <li
    className="pagination__item"
    data-testid="pagination-item-2"
    onClick={[Function]}
    value={2}
  >
    2
  </li>
  <li
    className="pagination__item"
    data-testid="pagination-item-3"
    onClick={[Function]}
    value={3}
  >
    3
  </li>
  <li
    className="pagination__item"
    data-testid="pagination-item-4"
    onClick={[Function]}
    value={4}
  >
    4
  </li>
  <li
    className="pagination__item pagination__item_active"
    data-testid="pagination-item-5"
    onClick={[Function]}
    value={5}
  >
    5
  </li>
  <li
    className="pagination__item pagination__item_disabled"
    data-testid="pagination-item-6"
    onClick={[Function]}
    value={6}
  >
    &gt;
  </li>
</ul>
`;

exports[`<Pagination /> Render Pagination with no data 1`] = `null`;
